### ISSUE TITLE
ROLL-49 : Service Class Revamp

### ISSUE THREAD

**Issue by brickgriff** [bgriffiniv/rollme#85]
Auth Service still handles requests by invoking callbacks during subscriptions and while this may work well it does not fit with the current direction of returning `Observables` to be used asynchronously by the caller.

**Comment by brickgriff** [comment]
Callbacks work great for certain scenarios. I propose the following interface for our three firebase services:

```
// note
var callback = (error, data) => {
  // do something ...
};

var success = callback(null, result);
var failure = callback(reason);
```

auth.service
```
  isAuthenticated(): boolean
  getCurrentUserId(): string

  getCurrentUser(callback) // necessary?

  signUp(email: string, password: string, callback)
  signIn(email: string, password: string, callback)
  signOut(callback)
```

card.service
```
  listCards(callback)
  listCardsByOwner(ownerId: string, callback)
  listCardsByHolder(holderId: string, callback)

  addCard(card: Card, callback) // maybe createCard(...)?
  getCard(id: string, callback) // maybe readCard(...)?

  updateCard(id: string, card: Card, callback)
  deleteCard(id: string, callback)

  listStaticCards(callback)
  listStaticCardsByOwner(ownerId: string, callback)
  listStaticCardsByHolder(holderId: string, callback)

  addStaticCard(card: Card, callback) // maybe createStaticCard(...)?
  getStaticCards(id: string, callback) // maybe readStaticCard(...)?

  updateStaticCard(id: string, card: Card, callback)
  deleteStaticCard(id: string, callback)
```

user.service
```
  listUsers(callback)

  addUser(user: User, callback) // ... create* ...
  getUser(id: string, callback) // ... read* ...

  updateUser(id: string, user: User, callback)
  deleteUser(id: string, callback)

  listStaticUsers(callback)

  addStaticUser(user: User, callback) // ... create* ...
  getStaticUser(id: string, callback) // ... read* ...

  updateStaticUser(id: string, user: User, callback)
  deleteStaticUser(id: string, callback)
```

**Comment by brandlxee11** [comment]
I agree. Good issue. Part two to this revamp in the next issue can focus on creating new services to support the other two app features: rolodex mechanics and exchange logistics. What do you think?

**Comment by brickgriff** [comment]
Note: I've made some pretty major changes, particularly to the `list()` functions. I've added an `id` field to the `update()` and `delete()` methods. This is because, although the `list()` functions will return objects with an `id` field, the objects returned by the `get()` functions will not know their own `id`. 

To put this another way, to avoid the complexity of adding an `id` to individual objects, only the collections should care about the `id`. This can be used to create links to the details pages but, once inside the details pages, only the details will be reliably available to the object while the `id` will be available via the URL (which can be used for updates and deletions).

So, moving forward, sending or expecting something like `card.id` is not going to work in services; we should not assume that an object knows its own `id`.
